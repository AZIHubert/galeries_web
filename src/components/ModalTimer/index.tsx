import * as React from 'react';
import * as ReactDOM from 'react-dom';

import { Container } from './styles';

type Variant = 'danger' | 'primary';

const modalRoot = document.getElementById('modal-root');
interface ModalTimerI {
  handleClose: () => void;
  open: boolean;
  testId?: string;
  text: string;
  variant?: Variant;
}

const ModalTimer = ({
  handleClose,
  open,
  testId,
  text,
  variant = 'primary',
}: ModalTimerI) => {
  const el = React.useRef(document.createElement('div'));
  const timer = React.useRef<ReturnType<typeof setTimeout> | null>(null);

  React.useEffect(() => {
    if (open) {
      timer.current = setTimeout(() => handleClose(), 3000);
      if (modalRoot) {
        modalRoot.appendChild(el.current);
      }
    }
    return () => {
      if (modalRoot && open) {
        modalRoot.removeChild(el.current);
      }
      if (timer.current) {
        clearTimeout(timer.current);
      }
    };
  }, [open]);
  if (!open) return null;
  return (
    ReactDOM.createPortal(
      <Container
        testId={testId}
        variant={variant}
      >
        {text}
      </Container>,
      el.current,
    )
  );
};

export default ModalTimer;
