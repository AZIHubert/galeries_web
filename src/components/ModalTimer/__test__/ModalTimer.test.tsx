import {
  act,
  render,
  screen,
} from '@testing-library/react';
import * as React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import ThemeProvider from '#contexts/ThemeContext';

import ModalTimer from '../index';

const Container = ({
  defaultOpen = false,
}: {
  defaultOpen?: boolean;
}) => {
  const [open, setOpen] = React.useState<boolean>(defaultOpen);
  const handleClose = () => setOpen(false);
  return (
    <ThemeProvider>
      <ModalTimer
        handleClose={handleClose}
        open={open}
        text='text'
        testId='modal'
      />
    </ThemeProvider>
  );
};

jest.useFakeTimers();

describe('Desktop', () => {
  const MockedHandleClose = jest.fn;
  beforeEach(() => {
  });
  beforeAll(() => {
    // @ts-ignore
    ReactDOM.createPortal = jest.fn((element) => element);
  });
  afterEach(() => {
    // @ts-ignore
    ReactDOM.createPortal.mockClear();
    // jest.runOnlyPendingTimers();
    // jest.useRealTimers();
  });
  it('renders without crashing', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <ModalTimer
          handleClose={MockedHandleClose}
          open={true}
          text='render text'
        />
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render if open is true', () => {
    act(() => {
      render(
        <Container
          defaultOpen={true}
        />,
      );
    });
    const modal = screen.queryByTestId('modal');
    expect(modal).not.toBeNull();
  });
  it('should not render if open is false', () => {
    act(() => {
      render(
        <Container />,
      );
    });
    const modal = screen.queryByTestId('modal');
    expect(modal).toBeNull();
  });
  it('should unmout after 3 second', () => {
    act(() => {
      render(
        <Container
          defaultOpen={true}
        />,
      );
    });
    act(() => {
      jest.advanceTimersByTime(1500);
    });
    const modalOpen = screen.queryByTestId('modal');
    expect(modalOpen).not.toBeNull();
    act(() => {
      jest.advanceTimersByTime(1500);
    });
    const modalClose = screen.queryByTestId('modal');
    expect(modalClose).toBeNull();
  });
});
