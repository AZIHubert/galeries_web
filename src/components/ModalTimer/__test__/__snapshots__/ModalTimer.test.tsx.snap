// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desktop renders without crashing 1`] = `
<div
  className="sc-bdfBwQ kLZtOh"
>
  render text
</div>
`;
